### PR TITLE
fix NPE in Supervisor.Stop when job has gracePeriod

### PR DESCRIPTION
**Pull request by mattnworb**
It's possible for a Supervisor to receive a goal of UNDEPLOY for a job that
it has not been told to START - perhaps because the agent was restarted
during the undeploy of a previous job (during the grace period sleeping).

Since the TaskRunner is only set after receiving the START goal, the call
to `runner.unregister()` when the job being undeployed throws a
NullPointerException.

Add a test to verify behavior.

### CODE REVIEW

File: helios-services/src/main/java/com/spotify/helios/agent/Supervisor.java [hunk 1]
```diff
@@ -349,14 +349,16 @@ public void perform(final boolean done) throws InterruptedException {
...
 
-      final Integer gracePeriod = job.getGracePeriod();
-      if (gracePeriod != null && gracePeriod > 0) {
-        log.info("Unregistering from service discovery for {} seconds before stopping",
-                 gracePeriod);
-
-        if (runner.unregister()) {
-          log.info("Unregistered. Now sleeping for {} seconds.", gracePeriod);
-          sleeper.sleep(TimeUnit.MILLISECONDS.convert(gracePeriod, TimeUnit.SECONDS));
+      if (runner != null) {
```
Comment (mattnworb) [inline helios-services/src/main/java/com/spotify/helios/agent/Supervisor.java]: here is the actual fix

File: helios-services/src/test/java/com/spotify/helios/agent/SupervisorTest.java [hunk 2]
```diff
@@ -396,12 +365,78 @@ public void verifyRuntimeExceptionSetsTaskStatusToFailed() throws Exception {
...
-    return new Answer<Object>() {
+    return (Answer<Object>) invocation -> future.get();
+  }
+
+  /**
+   * Verifies a fix for a NPE that is thrown when the Supervisor receives a goal of UNDEPLOY for a
+   * job with gracePeriod that has never been STARTed.
+   */
+  @Test
+  public void verifySupervisorHandlesUndeployingOfNotRunningContainerWithGracePeriod()
```
Comment (mattnworb) [inline helios-services/src/test/java/com/spotify/helios/agent/SupervisorTest.java]: and here is the actual test - the rest of the changes in SupervisorTest are refactoring to make this test possible (to use a different Job instance)

**Comment by mattnworb** [comment]
@spotify/helios-team

**Comment by gimaker** [comment]
:+1:

**Pull request merged**